utils/providers.py

```diff
@@ -1,388 +1,391 @@
-problematic_providers = set([])
 
+class ProviderFilter(object):
+    """ Filter used to obtain only providers matching given requirements
 
-def list_providers(allowed_types=None):
-    """ Returns list of providers of selected type from configuration.
+    Args:
+        keys: List of acceptable provider keys, all if `None`
+        categories: List of acceptable provider categories, all if `None`
+        types: List of acceptable provider types, all if `None`
+        required_fields: List of required fields,
+                         see :py:func:`providers_by_type`
+        restrict_version: Checks provider version in yamls if `True`
+        required_tags: List of tags that must be set in yamls
+        inverted: Inclusive if `False`, exclusive otherwise
+    """
+    _version_operator_map = OrderedDict([('>=', lambda o, v: o >= v),
+                                        ('<=', lambda o, v: o <= v),
+                                        ('==', lambda o, v: o == v),
+                                        ('!=', lambda o, v: o != v),
+                                        ('>', lambda o, v: o > v),
+                                        ('<', lambda o, v: o < v)])
+
+    def __init__(self, keys=None, categories=None, types=None, required_fields=None,
+                 restrict_version=True, required_tags=None, inverted=False):
+        self.keys = keys
+        self.categories = categories
+        self.types = types
+        self.required_fields = required_fields
+        self.restrict_version = restrict_version
+        self.required_tags = required_tags
+        self.inverted = inverted
+
+    def _filter_restricted_version(self, provider):
+        restricted_version = provider.data.get('restricted_version', None)
+        if restricted_version:
+            logger.info('we found a restricted version')
+            for op, comparator in ProviderFilter._version_operator_map.items():
+                # split string by op; if the split works, version won't be empty
+                head, op, ver = restricted_version.partition(op)
+                if not ver:  # This means that the operator was not found
+                    continue
+                if not comparator(version.current_version(), ver):
+                    return False
+                break
+            else:
+                raise Exception('Operator not found in {}'.format(restricted_version))
+        return True
 
-    @param allowed_types: Passed by partial(), see top of this file.
-    @type allowed_types: dict, list, set, tuple
+    def _filter_required_fields(self, provider, required_fields):
+        for field_or_fields in required_fields:
+            if isinstance(field_or_fields, tuple):
+                field_ident, field_value = field_or_fields
+            else:
+                field_ident, field_value = field_or_fields, None
+            if isinstance(field_ident, basestring):
```

Comment:
Since we already include the `six` library, you can prepare the code for py3 by using `six.string_types` instead of `basestring` which does not exist in py3.